### PR TITLE
[DTensor] Fix default_strategy and rename for clarity

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158538
* __->__ #158490

Fixes several bugs in the original.
- foremost, fixes a serious bug where we returned incorrect strategies
  by mixing input_specs that were frozen from
  select_strategy.strategies[0] with output_specs that varied across
  select_strategy.strategies[0..N] (e.g. we could create a nonsense
  strategy like input:Shard(0) output(Replicate) for an op like clone
- fixes the redistribute costs: they should not actually be 0, they
  should be the cost of redistributing our single input from another
  strategy to the current strategy, in our list of output strategies
- adds a note, wondering if we should have just literally returned the
  input strategy instead of creating this new object
- Currently, using default_strategy is incorrect becuase it maps 'self'
  tensor's strategies directly onto 'src' tensor without accounting for
  the fact that copy_ supports broadcasting a smaller rank tensor into a
  larger one.

Separates out copy_  op from default strategy, adds missing test case,
but does not fix the underlying issue with copy_, leaves that for future
PR

Renames to `propagate_single_input_strategy` since that's more
descriptive

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta

### CODE REVIEW

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 1]
```diff
@@ -82,12 +85,12 @@ def default_strategy(op_schema: OpSchema) -> StrategyType:
         aten.view.dtype,
         aten.zero_.default,
     ]
-)(default_strategy)
+)(propagate_single_input_strategy)
```
Comment (zpcore) [inline torch/distributed/tensor/_ops/_tensor_ops.py:87]: Doesn't look like all ops here only have one input arg, e.g., `aten.fill_.Scalar`.

Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py:87]: i was wondering about that. but fill_.scalar should still only have one 'tensor' input, so i think my statement still holds and I just need to find a better way to check the schema?

Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py:87]: trying this instead
`assert len([s for s in op_schema.args_schema if isinstance(s, OpStrategy)]) == 1`

Comment (zpcore) [reply torch/distributed/tensor/_ops/_tensor_ops.py:87]: Yes, one Tensor input should hold. LGTM now!

Comment (zpcore) [reply torch/distributed/tensor/_ops/_tensor_ops.py:87]: Oh, wait, `func: copy_(Tensor(a!) self, Tensor src, bool non_blocking=False) -> Tensor(a!)` still have two.

Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py:87]: you're right. hmm. i don't really like using this strategy func for an op that has 2 tensor inputs.  

the way i interpret that is, we actually have 2 tensor inputs, each with _different_ strategies.  Let's pick the first one arbitrarily, ignore the second one, and then create an op strategy where we use the first input's strategy in place of the second input.

Will this work for to_copy? I am suspicious that it only works in some cases.

Counter example:
If we feed an input tensor of shape (2,) into copy_, we can 'broadcast' it implicitly.

```
>>> import torch
>>> x = torch.ones([2,2])
>>> y = torch.ones([2,]) * 3
>>> x.copy_(y)
tensor([[3., 3.],
        [3., 3.]])
```

If 'self' tensor (our first tensor input) has a sharding on dim 1, we can NOT use that sharding to describe 'src' tensor, whihch only has dim0.  So it is not correct to use this strategy for copy_ in general.

Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py:87]: it also appears that we do not have a unit test for `copy_` in the first place, but i added one that confirms that the assert len strategies == 1 will fail and we need a fix.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 2]
```diff
@@ -39,55 +39,98 @@
...
-    # the tensor metas are distinct between the arguments and outputs
-    input_specs = []
-    redistribute_cost = []
-    for i in op_schema.args_schema:
-        input_specs.append(
-            DTensorSpec(
-                mesh=select_strategy.mesh,
-                placements=select_strategy.strategies[0].output_spec.placements,
-                tensor_meta=select_strategy.strategies[0].output_spec.tensor_meta,
+def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
```
Comment (wanchaol) [inline torch/distributed/tensor/_ops/_tensor_ops.py:42]: The `default_strategy` is not a single_input strategy, if you look at the old code, it actually loop through all the args_schema and append a new DTensorSpec for each argument, so it is meant be able to handle multiple input arguments.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 3]
```diff
@@ -39,55 +39,98 @@
...
+            OpSpec(
+                output_specs=DTensorSpec(
+                    mesh=first_input_strategy.mesh,
+                    placements=strategy.output_spec.placements,
+                    tensor_meta=strategy.output_spec.tensor_meta,
+                ),
+                input_specs=[
+                    DTensorSpec(
+                        mesh=first_input_strategy.mesh,
+                        placements=strategy.output_spec.placements,
```
Comment (wanchaol) [inline torch/distributed/tensor/_ops/_tensor_ops.py:63]: The intention of default strategy means to let every argument follow the input argument strategy, but here it looks like you only assume the op have a single input. Probably make sense for the renaming you have, but it was not the same intended behavior as before

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 4]
```diff
@@ -39,55 +39,98 @@
...
-
-    default_strategy = [
-        OpSpec(
-            output_specs=DTensorSpec(
-                mesh=select_strategy.mesh,
-                placements=strategy.output_spec.placements,
-                tensor_meta=strategy.output_spec.tensor_meta,
-            ),
-            input_specs=input_specs,
-            redistribute_cost=redistribute_cost,
```
Comment (wanchaol) [inline torch/distributed/tensor/_ops/_tensor_ops.py:68]: aren't the real fix here should simply be generating the redistribute_cost here by calling                `generate_redistribute_costs`?

**Review by wanchaol — CHANGES_REQUESTED** [review]
Could you explain more about the bug? I thought it could be easily fixed by dynamically generate redistribute_costs. This changes the intention of default_strategy and I wonder if it's necessary

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 3 above)
Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py:63]: I found that out of all the ops that we used 'default_strategy' for, only `copy_` was "benefitting" from applying the first input strategy to the second input.

however, for `copy_`, this supposed benefit was actually a hidden correctness bug.  It works for cases where src,self are same dim, but it fails for cases where src is smaller dim than self and requires broadcasting.  I added a new test case that will fail with the current `default_strategy` applied to copy.

Based on this, and since we do not have an example of an op that has multiple inputs that can safely share the first input's strategy, I decided its better to just simplify the helper to a narrower purpose.  Wdyt?

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 4 above)
Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py:68]: this fixes only the second bug in my PR desc.  I'll explain the first bug more in a comment below.

**Comment by wconstab** [comment]
> Could you explain more about the bug? I thought it could be easily fixed by dynamically generate redistribute_costs. This changes the intention of default_strategy and I wonder if it's necessary

Yes, the first bug in my PR desc is related to the input specs not the redistribute costs.

> (e.g. we could create a nonsense strategy like input:Shard(0) output(Replicate) for an op like clone

i found empirically that we were creating a strategy for Clone() where input strategy was Shard(0) and output strategy was Replicate().  I realized the reason was that we were for-looping over input strategies but hardcoding output strategy to be the one from the first input strategy.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 5]
```diff
@@ -39,55 +39,98 @@
...
-        input_specs.append(
-            DTensorSpec(
-                mesh=select_strategy.mesh,
-                placements=select_strategy.strategies[0].output_spec.placements,
-                tensor_meta=select_strategy.strategies[0].output_spec.tensor_meta,
+def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
+    # For ops with a single tensor input, we perform a 1:1 mapping such that
+    # for each strategy that the input supports, we create a corresponding strategy.
+    # Note: this may be a complete waste of work, becuase it should be equivalent to
+    # `return first_input_strategy` (unless creating a deep copy is important for some reason)
```
Comment (XilunWu) [inline torch/distributed/tensor/_ops/_tensor_ops.py]: I don't think we're modifying `OpSpec` anywhere so it should be safe to directly return.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 6]
```diff
@@ -39,55 +39,98 @@
...
+                redistribute_cost=[
+                    generate_redistribute_costs(
+                        first_input_strategy, strategy.output_spec
+                    )
+                    for _ in range(num_tensor_args)
+                ],
+            )
+            for strategy in first_input_strategy.strategies
+        ]
+    )
```
Comment (XilunWu) [inline torch/distributed/tensor/_ops/_tensor_ops.py:133]: IIUC this doesn't address the broadcast issue either. Do we address it in another PR?

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 2 above)
Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py:42]: yes, i realized i was changing the semantic of the function.  However, it was not clear if the original design was fully thought-through.  5/6 ops using it were single-input, and the other op that has multi inputs was using it incorrectly by accident.  

Can you convince me that there are ops that we want to use the multi-input default strategy for and would be correct?  If not, lets just make it simpler, and avoid the possibility of using it incorrectly by accident.

**Review by XilunWu — COMMENTED** [review]
The idea that we need a `single_operand_strategy` looks good to me. Would like discuss more:

The pattern of "all DTensor input should follow a given input's sharding" seems common. I believe we still need it (of course it needs fix) with a new name. But I'm okay with removing it in this PR and add a correct version back in another since the only use case right now is `copy_` and `copy_` needs to handle broadcast. 

BTW, do some `foreach_*` ops need single-input propagate and follow-some-input propagate as well?

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 6 above)
Comment (wconstab) [reply torch/distributed/tensor/_ops/_tensor_ops.py:133]: see the next PR in the stack

**Comment by wconstab** [comment]
> BTW, do some foreach_* ops need single-input propagate and follow-some-input propagate as well?

foreach ops are different in that they are many to many, not many to one.  the previous default_strategy was many to one.  (one output).  It would be bad to use 'default_strategy' for foreach, becuase each output needs to follow _its own_ input, not the first input.

**Comment by wconstab** [comment]
> The pattern of "all DTensor input should follow a given input's sharding" seems common. I believe we still need it (of course it needs fix) with a new name. But I'm okay with removing it in this PR and add a correct version back in another since the only use case right now is copy_ and copy_ needs to handle broadcast.

can you give an example of an op where this pattern is true?

File: test/distributed/tensor/test_tensor_ops.py [hunk 7]
```diff
@@ -53,6 +53,38 @@ def test_clone(self):
...
+    #     dst_tensor = torch.zeros(12, 12)
+    #     dst_specs = [[Replicate()], [Shard(1)]]
+    #     for dst_spec, src_spec in zip(dst_specs, src_specs):
+    #         src_dtensor = distribute_tensor(src_tensor, device_mesh, dst_spec)
+    #         dst_dtensor = distribute_tensor(dst_tensor, device_mesh, src_spec)
+    #         # perform a broadcasted copy from Shard(0) to Shard(1) for the worst case
+    #         dst_dtensor.copy_(src_dtensor)
+    #         dst_tensor.copy_(src_tensor)
+    #         self.assertEqual(dst_dtensor.full_tensor(), dst_tensor)
+
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py:87]: let's remove this

**Review by XilunWu — APPROVED** [review]
The strategy part LGTM. Didn't review the `copy_` part because they're override in next PR.

File: test/distributed/tensor/test_tensor_ops.py [hunk 8]
```diff
@@ -53,6 +53,38 @@ def test_clone(self):
...
+        for dst_spec, src_spec in zip(dst_specs, src_specs):
+            src_dtensor = distribute_tensor(src_tensor, device_mesh, dst_spec)
+            dst_dtensor = distribute_tensor(dst_tensor, device_mesh, src_spec)
+            dst_dtensor.copy_(src_dtensor)
+            dst_tensor.copy_(src_tensor)
+            self.assertEqual(dst_dtensor.full_tensor(), dst_tensor)
+
+    # @pytest.mark.xfail
+    # @with_comms
+    # def test_copy_broadcast(self):
```
Comment (wanchaol) [inline test/distributed/tensor/test_tensor_ops.py:73]: why this being comment out? I thought you are marking xfails

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/tensor/_ops/_tensor_ops.py:42]: Oh actually i think it's probably fine, it was added in this PR https://github.com/pytorch/pytorch/commit/ae86e8f6c829a3cfa9204949156fce2d048c919b

Before that default strategy does not handle multiple inputs anyways.

**Review by wanchaol — APPROVED** [review]

File: test/distributed/tensor/test_tensor_ops.py (hunk 8 above)
Comment (wconstab) [reply test/distributed/tensor/test_tensor_ops.py:73]: xfail did not work haha. so i commented it. its uncommented and passing in the next PR.

**Comment by clee2000** [comment]
@pytorchbot revert -m "broke lint? [GH job link](https://github.com/pytorch/pytorch/actions/runs/16361950974/job/46231492581) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/d8b084312b54e97bdbaf6a178fe2fc628a23243b)" -c landrace

cc @seemethere tentatively marking this as landrace but the misspelling does exist in the PR as well
